src/der.rs

```diff
@@ -117,26 +125,26 @@ fn nonnegative_integer<'a>(input: &mut untrusted::Reader<'a>, min_value: u8)
 
     let value = try!(expect_tag_and_get_value(input, Tag::Integer));
 
-    value.read_all(error::Unspecified, |input| {
+    value.read_all(ASN1Error, |input| {
```

Comment:
> Since the point of this PR is to allow us to differentiate different kinds of errors, it seems like we should differentiate out-of-range vs. bad-DER. WDYT?

I would love to; the only reason I didn't do so was to hew more closely to the "expose as little as possible" goal. More on this below.

> The end-goal of this specific PR is to make it possible to rebase the webpki crate on top of the new version of ring that has #378 integrated, right? 

It would be possible, though a trifle code-smelly, to use this (plus #378) in webpki. The issue is the, ahem, *approximate* error handling like this, which is currently residing in #378:

```
            try!(der::expect_tag_and_get_value(input, der::Tag::Sequence)
                .map_err(|_| ParseSPKIError::BadDER)); 
```

It would be nice to not map `Unspecified` to `BadDER` just because it seems like that's probably what's going on... and sure enough, webpki has its own way of doing this:

```
#[inline(always)]
pub fn expect_tag_and_get_value<'a>(input: &mut untrusted::Reader<'a>,
                                    tag: Tag) ->
                                    Result<untrusted::Input<'a>, Error> {
    ring::der::expect_tag_and_get_value(input, tag).map_err(|_| Error::BadDER)
}
```

Mapping `Unspecified` to `BadDER` may be correct, but it doesn't feel very good.

Anyway, consider webpki's `parse_spki_value`:

```
fn parse_spki_value(input: untrusted::Input)
                    -> Result<SubjectPublicKeyInfo, Error> {
    input.read_all(Error::BadDER, |input| {
        let algorithm_id_value =
                try!(der::expect_tag_and_get_value(input, der::Tag::Sequence));
        let key_value = try!(der::bit_string_with_no_unused_bits(input));
...
```

In practice, the only `Error` variant this will return is `BadDER`. So, when webpki uses *ring* for its spki parsing needs and effectively replaces that function, it could quite neatly map `ParseSPKIError::BadDER` to `Error::BadDER` and call it a day, with or without a better breakdown of ASN.1 errors that would presumably all get mapped to `ParseSPKIError::BadDER` (maybe with a field in that variant to store which specific `ASN1Error` it was?).

In summary: no, I don't think we need more detailed errors to make it possible to use this in webpki. However, I wholeheartedly agree that it is not a very nice state of the world to report an value-out-of-bounds error the same way as an unexpected EOF, so I would happily make `ASN1Error` an enum with `UnexpectedEOF`, `ValueOutOfBounds`, etc variants if you're OK with that (or some other error scheme, though that's what came to mind). Based on your other comments, it sounds like parsing ASN.1 is not itself considered sensitive, so more detailed information about its parsing should not be a problem, and you know how much I love building to allow testability... :)